File: ruby/nucleotide-count/nucleotide_count.rb
class Nucleotide
  BASES = ["A", "T", "C", "G"]

  def self.from_dna(dna)
    raise ArgumentError unless valid?(dna)
    Nucleotide.new(dna)
  end

  def self.valid?(dna)
    dna.delete(BASES.join).empty?
  end

  attr_accessor :dna

  def initialize(dna)
    @dna = dna
  end

  def count(base)
    dna.count(base)
  end

  def histogram
    BASES.map { |base| [base, count(base)] }.to_h
  end
end

File: elixir/nucleotide-count/nucleotide_count.exs
defmodule NucleotideCount do
  @nucleotides [?A, ?C, ?G, ?T]

  def count(dna, base) do
    Enum.count(dna, fn x -> x == base end)
  end

  def histogram(dna) do
    Map.new(@nucleotides, fn base -> { base, count(dna, base)} end)
  end
end
